pyomo/contrib/doe/fim_doe.py

```diff
@@ -0,0 +1,2582 @@
+            3.  sequential_kaug: calculate sensitivity by k_aug [Experimental]
+            4.  direct_kaug: calculate sensitivity by k_aug with direct sensitivity
+
+        Parameters
+        -----------
+        design_values:
+            a ``dict`` where keys are design variable names, values are a dict whose keys are time point and values are the design variable value at that time point
+        design_ranges:
+            a ``list`` of design variable values to go over
+        design_dimension_names:
+            a ``list`` of design variable names of each design range
+        design_control_time:
+            a ``list`` of control time points that should be fixed to the values in dv_ranges
+        mode:
+            use mode='sequential_finite', 'sequential_sipopt', 'sequential_kaug', 'direct_kaug'
+        tee_option:
+            if solver console output is made
+        scale_nominal_param_value:
+            if True, the parameters are scaled by its own nominal value in param_init
+        scale_constant_value:
+            scale all elements in Jacobian matrix, default is 1.
+        store_name:
+            a string of file name. If not None, store results with this name.
+            Since there are maultiple experiments, results are numbered with a scalar number, 
+            and the result for one grid is 'store_name(count).csv' (count is the number of count).
+        read_name: 
+            a string of file name. If not None, read result files. 
+            Since there are multiple experiments, this string should be the common part of all files.
+        filename:
+            if True, grid search results stored with this file name
+        formula:
+            choose from 'central', 'forward', 'backward', None. This option is only used for 'sequential_finite' mode.
+        step:
+            Sensitivity perturbation step size, a fraction between [0,1]. default is 0.001
+
+        Return
+        -------
+        figure_draw_object: a combined result object of class Grid_search_result
+        """
+
+        # time 0
+        t_enumeration_begin = time.time()
+
+        # Set the Objective Function to 0 helps solve square problem quickly
+        self.objective_option='zero'
+        self.filename = filename
+
+        # calculate how much the FIM element is scaled
+        self.fim_scale_constant_value = scale_constant_value ** 2
+
+        # when defining design space, design variable values are defined as in design_values argument
+        # the design var value defined in dv_ranges only applies to control time points given in dv_apply_time
+        grid_dimension = len(design_ranges)
+
+        # to store all FIM results
+        result_combine = {}
+
+        # iteration 0
+        count = 0
+        failed_count = 0
+        # how many sets of design variables will be run
+        total_count = 1
+        for i in range(grid_dimension):
+            total_count *= len(design_ranges[i])
```

Comment:
It is both simpler and faster to just iterate over the list instead of iterating over the list indices and extracting each element:
```python
total_count = 1
for rng in design_ranges:
    total_count *= len(rng)
```
(you can make similar changes throughout this file.)